crates/ra_hir/src/ty/pretty.rs

```diff
@@ -0,0 +1,48 @@
+use std::fmt;
+
+use crate::db::HirDatabase;
+
+pub struct PrettyPrintDisplay<'a, DB, T>(&'a DB, &'a T);
+
+impl<'a, DB, T> fmt::Display for PrettyPrintDisplay<'a, DB, T>
+where
+    DB: HirDatabase,
+    T: PrettyPrint,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.1.pp(self.0, f)
+    }
+}
+
+pub trait PrettyPrint {
+    fn pp(&self, db: &impl HirDatabase, f: &mut fmt::Formatter) -> fmt::Result;
```

Comment:
Instead of passing a pair, we can pass a HIrFormatter with db and fmt fields. This should help with recursive impls (only one arg to pass) and would be a good place to add join method.